### PR TITLE
Init ResourceUsage to prevent invalid values

### PR DESCRIPTION
**Pull request by Flakebi**
In some cases, fields stay uninitialized but are read.
This bug was found by ubsan.

### CODE REVIEW

**Comment by Flakebi** [comment]
Thinking about the code, all the code in `InitShaderResourceUsage` initializes a `ResourceUsage` to its default values. Shouldn’t we put that into a constructor of `ResourceUsage(ShaderStage)` then?
This sounds a lot cleaner to me.

**Comment by trenouf** [comment]
Yes, sounds good to me.

File: lgc/CMakeLists.txt [hunk 1]
```diff
@@ -235,4 +236,3 @@ target_sources(LLVMlgc PRIVATE
     util/llpcPipelineShaders.cpp
     util/llpcStartStopTimer.cpp
 )
```
Comment (kuhar) [inline lgc/CMakeLists.txt:238]: nit: unintended?

File: lgc/builder/llpcResourceUsage.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+#include "llpcResourceUsage.h"
+
+using namespace lgc;
+
+ResourceUsage::ResourceUsage(ShaderStage shaderStage)
+{
+    memset(&builtInUsage, 0, sizeof(builtInUsage));
+    memset(inOutUsage.gs.outLocCount, 0, sizeof(inOutUsage.gs.outLocCount));
+    memset(inOutUsage.xfbStrides, 0, sizeof(inOutUsage.xfbStrides));
+    memset(inOutUsage.streamXfbBuffers, 0, sizeof(inOutUsage.streamXfbBuffers));
```
Comment (kuhar) [inline lgc/builder/llpcResourceUsage.cpp:40]: could (some of) these be replaced with `= {}` in the class definition?

File: lgc/builder/llpcResourceUsage.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    }
+}
+
+InterfaceData::InterfaceData()
+{
+    memset(userDataMap, InterfaceData::UserDataUnmapped, sizeof(userDataMap));
+    memset(&pushConst, 0, sizeof(pushConst));
+    memset(&spillTable, 0, sizeof(spillTable));
+    memset(&userDataUsage, 0, sizeof(userDataUsage));
+    memset(&entryArgIdxs, 0, sizeof(entryArgIdxs));
```
Comment (kuhar) [inline lgc/builder/llpcResourceUsage.cpp:90]: Same here

File: lgc/builder/llpcResourceUsage.h [hunk 4]
```diff
@@ -474,21 +478,21 @@ struct InterfaceData
...
     {
-        uint32_t                resNodeIdx;                       // Resource node index for push constant
+        uint32_t          resNodeIdx = InvalidValue;        // Resource node index for push constant
     } pushConst;
 
     struct
     {
-        uint32_t                sizeInDwords;                     // Spill table size in dwords
-        uint32_t                offsetInDwords;                   // Start offset of Spill table
+        uint32_t          sizeInDwords;                     // Spill table size in dwords
```
Comment (kuhar) [inline lgc/builder/llpcResourceUsage.h:494]: missed this one?

File: lgc/builder/llpcResourceUsage.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+ ***********************************************************************************************************************
+ * @file  llpcResourceUsage.cpp
+ * @brief LLPC source file: contains implementation of ResourceUsage and InterfaceData.
+ ***********************************************************************************************************************
+ */
+#include "llpcResourceUsage.h"
+
+using namespace lgc;
+
+ResourceUsage::ResourceUsage(ShaderStage shaderStage)
```
Comment (amdrexu) [inline lgc/builder/llpcResourceUsage.cpp:35]: Please add separate line
// ===
ResourceUsage::ResourceUsage(
    ShaderStage shaderStage) // Shader stage

File: lgc/builder/llpcResourceUsage.cpp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            inOutUsage.fs.outputTypes[i] = BasicType::Unknown;
+        }
+
+        inOutUsage.fs.cbShaderMask = 0;
+        inOutUsage.fs.dummyExport = true;
+        inOutUsage.fs.isNullFs = false;
+    }
+}
+
+InterfaceData::InterfaceData()
```
Comment (amdrexu) [inline lgc/builder/llpcResourceUsage.cpp:77]: Add separate line
// === (120 character)

**Comment by JaxLinAMD** [comment]
please resolve conflicts

File: lgc/builder/llpcResourceUsage.cpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+ * @file  llpcResourceUsage.cpp
+ * @brief LLPC source file: contains implementation of ResourceUsage and InterfaceData.
+ ***********************************************************************************************************************
+ */
+#include "llpcResourceUsage.h"
+
+using namespace lgc;
+
+// =====================================================================================================================
+ResourceUsage::ResourceUsage(ShaderStage shaderStage) {
```
Comment (amdrexu) [inline lgc/builder/llpcResourceUsage.cpp:36]: @trenouf This is not an issue. Just raise an interesting question. I find PAL style doesn't require function comments for constructors and destructors (but parameter comments are required if existing). Now, we switch to LLVM style and still require parameter comments. It looks somewhat strange in function header we don't have function description comments but have parameter description comments for some class constructors. Like this:
```
// =====
//
// @param pipelineState : Pipeline state
// @param pipelineShaders : API shaders in the pipeline
ShaderMerger::ShaderMerger(...)
```

**Comment by amdvlk-admin** [comment]
### Test summary for commit 5cf77fefd76e8a9f54d11bd82a66aa826912cda1
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev bb7398b1c7509a30ba4f7d947b5ad5549a4d58db</li>
<li>SPVGEN: dev dc2b146de03ef54f188172d8fa2f9be708cbfaca</li>
<li>XGL: dev ffd39443fec0ce821b7702173eba7184855c7b5e</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 7b14353794b408b8d7fd87ae9a9f4f553e6d4782</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 955/955 (100.0%)</li>
<li>Failed: 0/955 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103142/103142 (100.0%)</li>
<li>Failed: 0/103142 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 88462/88463 (100.0%)</li>
<li><details><summary>Failed: 1/88463 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.subgroups.arithmetic.compute.subgroupinclusiveadd_int16_t
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
Hm.. please resolve conflicts....

**Comment by JaxLinAMD** [comment]
conflict again

**Comment by amdvlk-admin** [comment]
### Test summary for commit d1c2046e342cea753c71e0d06281c566f08df4c2
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev bb7398b1c7509a30ba4f7d947b5ad5549a4d58db</li>
<li>SPVGEN: dev dc2b146de03ef54f188172d8fa2f9be708cbfaca</li>
<li>XGL: dev ffd39443fec0ce821b7702173eba7184855c7b5e</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 7b14353794b408b8d7fd87ae9a9f4f553e6d4782</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 955/955 (100.0%)</li>
<li>Failed: 0/955 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103142/103142 (100.0%)</li>
<li>Failed: 0/103142 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 38363/38364 (100.0%)</li>
<li><details><summary>Failed: 1/38364 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.spirv_assembly.instruction.graphics.16bit_storage.uniform_float_64_to_16.uniform_buffer_block_scalar_float_rtz_vert
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103198/103198 (100.0%)</li>
<li>Failed: 0/103198 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
merge process is started, please don't force-push any more.

**Comment by Flakebi** [comment]
Thanks!

**Pull request merged**